### ISSUE TITLE
Invalid SQL for "WHERE ... IN" with parentheses

### ISSUE THREAD

**Issue by commonquail** [DapperLib/Dapper#1819]
Using Dapper 2.0.90 and 

```
> select @@version;
Microsoft SQL Azure (RTM) - 12.0.2000.8 
	May 12 2022 23:11:24 
	Copyright (C) 2022 Microsoft Corporation
```

Given

```sql
SELECT 1 FROM foo WHERE bar IN (@ids)
```

Dapper produces either

```sql
SELECT 1 FROM foo WHERE bar IN (@ids)
```

or

```sql
SELECT 1 FROM foo WHERE bar IN ((@ids1,@ids2))
```

when the value for `@ids` is `new int[] { 1 }` respectively `new int[] { 1, 2 }`; that is, when the count exceeds 1. The second case produces the error

> Microsoft.Data.SqlClient.SqlException : Incorrect syntax near ','.

because SQL Server does not recognize `... IN ( ( ... , ... ) )`. Neither does MySQL 8. Postgres 14 also errors in this case but that's because it lacks an equality operator between integers and its non-standard tuple extension (`... (bar, bar) IN ((@ids1, @ids2))` should work).

The inconsistency in Dapper's behaviour _may_ be explained by Dapper in actuality doing nothing when the count does not exceed 1.

I would argue that Dapper's behaviour in this case should be strictly consistent, and that its inconsistency is an error.

Whether Dapper should support the parenthesized form in the first place or exclusively the unparenthesized form (as demonstrated in the README) seems to me more of a stylistic decisions. Personally, I would have said that Dapper should never have inserted the parentheses because those patently are not part of the parameter(s), however, obviously Dapper cannot now remove that form. Whether Dapper should preserve support for the parenthesized form is less clear; such code works purely by coincidence. Perhaps Dapper could learn to detect existing parentheses and transparently overwrite them (i.e. keep generating parentheses but avoiding nesting), however, this is a substantially more complex solution to build and I don't know if that would be backwards compatible with Postgres' tuples.

**Comment by mgravell** [comment]
Not at a PC, but: have you tried ... in @ids

? i.e. no parentheses

**Comment by commonquail** [comment]
Yes; to be clear, the example demonstrated in the README works as advertised.

**Comment by mgravell** [comment]
Then... use that? I'm not sure there's any claim that the other syntax will
do anything here

**Comment by commonquail** [comment]
Neither is there any warning that it will silently generate invalid SQL, but only in some circumstances.

**Comment by mgravell** [comment]
I haven't checked, but I suspect that maybe it is generating `in ***@***.***))`
in the first case, which has a redundant set of parents which are silently
ignored. So: it is always working consistently, just: sometimes that
consistency results in something that accidentally works.

**Comment by Bosegeek** [comment]
I think it is designed that way, in case of multiple params use it without the parantheses. Check here - [PassIntArray()](https://github.com/DapperLib/Dapper/blob/a31dfd3dd4d7f3f2580bd33c877199d7ef3e3ef9/tests/Dapper.Tests/ParameterTests.cs#L203)

**Comment by Rudde** [comment]
We have experienced a lot of error like `Incorrect syntax near ''@ids''.` When using syntax proposed in the comments (and tests) here `SELECT 1 FROM foo WHERE bar IN @ids`

But the problem is not consistent. It often just happens when the app is running, and doesn't recover. It can happen at startup, or it can happen while it's running. To solve the issue we often just have to restart the service, no code changes at all, the same binary. 

The type I pass in for `@ids` is an `IEnumerable<int>`, but I have tried `int[]` and `List<int>` too without much change.

**Comment by mgravell** [comment]
@Rudde  whatever that is, it is a different scenario - I would need to see a concrete example to offer an opinion
